services/analyze-sequelize-tables.js

```diff
@@ -62,100 +62,197 @@ function hasTimestamps(fields) {
+
+  return [schema, foreignKeys, primaryKeys, []];
+}
+
+// NOTICE: Check the foreign key's reference unicity
+function checkRefUnicity(table, columnName) {
+  const isPrimary = table[2].includes(columnName);
+  const isUnique = _.find(table[1], { column_name: columnName, column_type: 'UNIQUE' }) !== undefined
+    || _.find(table[1], { unique_indexes: columnName, column_type: 'PRIMARY KEY' }) !== undefined;
+  return isPrimary || isUnique;
+}
+
+// NOTICE: Format the references depending on the type of the association
+function setReference(fk, association, manyToManyFk) {
+  let reference = {};
+  if (association === 'belongsTo') {
+    reference = {
+      isBelongsTo: true,
+      ref: fk.foreign_table_name,
+      foreignKey: fk.column_name,
+      foreignKeyName: _.camelCase(fk.column_name),
+      association,
+    };
+  } else if (association !== 'belongsToMany') {
+    reference = {
+      isHasOneOrHasMany: true,
+      ref: fk.table_name,
+      foreignKey: fk.column_name,
+      foreignKeyName: _.camelCase(fk.column_name),
+      association,
+    };
+  } else {
+    reference = {
+      isBelongsToMany: true,
+      ref: fk.table_name,
+      foreignKey: fk.column_name,
+      otherKey: manyToManyFk.column_name,
+      association,
+      junctionTable: manyToManyFk.foreign_table_name,
+    };
+  }
+
+  // NOTICE: If the foreign key name and alias are the same, Sequelize will crash, we need
+  //         to handle this specific scenario generating a different foreign key name.
+  if (reference.foreignKeyName !== undefined && reference.foreignKeyName === reference.as) {
+    reference.foreignKeyName = `${reference.foreignKeyName}Key`;
+  }
+
+  if (fk.foreign_column_name !== 'id') {
+    reference.targetKey = fk.foreign_column_name;
+  }
+
+  return reference;
+}
+
+function getData(table, config) {
   return P
     .resolve(analyzeFields(table, config))
     .then((schema) => P.all([
       schema,
       tableForeignKeysAnalyzer.perform(table),
       analyzePrimaryKeys(schema),
     ]))
-    .spread(async (schema, foreignKeys, primaryKeys) => {
-      const fields = [];
-      const references = [];
+    .then((data) => checkFkUnicity(data));
+}
 
-      await P.each(Object.keys(schema), async (nameColumn) => {
-        const columnInfo = schema[nameColumn];
-        const type = await columnTypeGetter.perform(columnInfo, nameColumn, table);
-        const foreignKey = _.find(foreignKeys, { column_name: nameColumn });
+// NOTICE: Use the foreign key and reference properties to determine the associations
+//         and push them as references of the table.
+async function setAssociationType(aggregatedData) {
+  await P.mapSeries(Object.values(aggregatedData), async (table) => {
+    table[1].forEach((fk) => {
+      if (fk.column_type === 'FOREIGN KEY') {
+        let manyToMany = false;
+        if (fk.isInCompositeKey) {
+          const arrayUniqueIndexes = [table[2]];
+          arrayUniqueIndexes.forEach((uniqueIndexes) => {
```

Comment:
This makes no sense to me. You are doing a foreach on an array with only one element that you created the line before ?!

Reply:
It was using unique_indexes before too, but the belongsToMany wasn't strict enough, I forgot to update it